x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/FileTokensTool.java

```diff
@@ -92,41 +86,92 @@ protected void execute(Terminal terminal, OptionSet options, Environment env) th
+            final String principal = tuple.v1();
+            final String tokenName = tuple.v2();
+            final String qualifiedName = principal + "/" + tokenName;
+            final Path serviceTokensFile = FileServiceAccountsTokenStore.resolveFile(env);
+
+            FileAttributesChecker attributesChecker = new FileAttributesChecker(serviceTokensFile);
+            final Map<String, char[]> tokenHashes = new TreeMap<>(FileServiceAccountsTokenStore.parseFile(serviceTokensFile, null));
+
+            if (tokenHashes.remove(qualifiedName) == null) {
+                throw new UserException(ExitCodes.CODE_ERROR, "Service token [" + qualifiedName + "] does not exist");
+            } else {
+                FileServiceAccountsTokenStore.writeFile(serviceTokensFile, tokenHashes);
+            }
+            attributesChecker.check(terminal);
         }
     }
 
     static class ListFileTokenCommand extends EnvironmentAwareCommand {
 
+        private final OptionSpec<String> arguments;
+
         ListFileTokenCommand() {
             super("List file tokens for the specified service account");
+            this.arguments = parser.nonOptions("service-account-principal");
         }
 
         @Override
         protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
-            throw new UnsupportedOperationException("list command not implemented yet");
+            final List<String> args = arguments.values(options);
+            if (args.size() > 1) {
+                throw new UserException(
+                    ExitCodes.USAGE,
+                    "Expected at most one argument, service-account-principal, found extra: " + args.toString());
+            }
+            Predicate<String> filter = k -> true;
+            if (args.size() == 1) {
+                final String principal = args.get(0);
+                if (false == ServiceAccountService.isServiceAccountPrincipal(principal)) {
+                    throw new UserException(ExitCodes.NO_USER, "Unknown service account principal: [" + principal + "]. Must be one of ["
+                        + Strings.collectionToDelimitedString(ServiceAccountService.getServiceAccountPrincipals(), ",") + "]");
+                }
+                filter = filter.and(k -> k.startsWith(principal + "/"));
+            }
+            final Path serviceTokensFile = FileServiceAccountsTokenStore.resolveFile(env);
+            final Map<String, char[]> tokenHashes = new TreeMap<>(FileServiceAccountsTokenStore.parseFile(serviceTokensFile, null));
+            for (String key : tokenHashes.keySet()) {
+                if (filter.test(key)) {
+                    terminal.println(key);
+                }
+            }
+        }
+    }
+
+    static Tuple<String, String> parsePrincipalAndTokenName(List<String> arguments, Settings settings) throws UserException {
```

Comment:
Personal opinion: I find use of `Tuple` to be a sign that there's a missing class. Sometimes it's just a case of "I need to return 2 things", but more often than not, those two things have some relationship between them, and that could be expressed in a real class.

I think I had a comment on another issue where we might need something like a `ServiceTokenId` class. Maybe this is another sign of that.

Reply:
Makes sense. I think I have experienced the friction along the way as well. It is pretty minor so I can live with it. But formalise the concept is more future proof.